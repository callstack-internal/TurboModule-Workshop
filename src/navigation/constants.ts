export const Screens = {
  Main: {
    title: 'Catálogo de Perros',
    name: 'Main',
  },
  HowItWorks: {
    title: 'Cómo funciona',
    name: 'HowItWorks',
  },
  Home: {
    title: 'Home',
    name: 'Home',
  },
  DogDetails: {
    title: 'Detalles',
    name: 'DogDetails',
  },
  Adoptar: {
    title: 'Adoptar',
    name: 'Adoptar',
  },
  MisPerros: {
    title: 'Mis Perros',
    name: 'MisPerros',
  },
} as const;
