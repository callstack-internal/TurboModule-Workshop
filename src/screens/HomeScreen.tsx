import React from 'react';
import { StyleSheet, View } from 'react-native';
import DogList from '../components/DogList';

function HomeScreen() {
  return (
    <View style={styles.container}>
      <DogList />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#f8f9fa',
  },
});

export default HomeScreen;
