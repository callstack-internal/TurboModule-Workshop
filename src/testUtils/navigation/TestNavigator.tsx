import React, { useEffect } from 'react';

import {
  createNavigatorFactory,
  useNavigationBuilder,
  StackRouter,
  TabRouter,
  DrawerRouter,
  DefaultRouterOptions,
  StackRouterOptions,
  TabRouterOptions,
  DrawerRouterOptions,
  NavigationHelpers,
  ParamListBase,
  RouterFactory,
  Router,
  NavigationState,
  NavigationAction,
  NavigationContainer,
} from '@react-navigation/native';
import type {
  StackNavigationState,
  TabNavigationState,
  DrawerNavigationState,
} from '@react-navigation/native';
import type {
  StackActionHelpers,
  TabActionHelpers,
  DrawerActionHelpers,
} from '@react-navigation/native';
import { Text } from 'react-native';

type TestScreenOptions = Record<string, unknown>;

type NavigatorType = 'stack' | 'tabs' | 'drawer';

interface TestNavigatorImplProps extends DefaultRouterOptions {
  type: NavigatorType;
  screenOptions?: TestScreenOptions;
  initialRouteName?: string;
  children: React.ReactNode;
}

type AnyNavigationState =
  | StackNavigationState<ParamListBase>
  | TabNavigationState<ParamListBase>
  | DrawerNavigationState<ParamListBase>;

type AnyRouterOptions = StackRouterOptions | TabRouterOptions | DrawerRouterOptions;

type AnyActionHelpers =
  | StackActionHelpers<ParamListBase>
  | TabActionHelpers<ParamListBase>
  | DrawerActionHelpers<ParamListBase>;

type ExtractedAction = Exclude<
  Parameters<Router<NavigationState, NavigationAction>['getStateForAction']>[0],
  undefined
>;

/**
 * Custom navigator component used for testing React Navigation screens.
 * Supports `stack`, `tabs`, and `drawer` types.
 * Renders all screens at once and injects a `focused` prop based on navigation state.
 *
 * Useful in tests to simulate navigation behavior with a lightweight navigator.
 */
const TestNavigatorImpl = ({ type, ...rest }: TestNavigatorImplProps): React.ReactElement => {
  const router = (
    type === 'drawer' ? DrawerRouter : type === 'stack' ? StackRouter : TabRouter
  ) as RouterFactory<AnyNavigationState, ExtractedAction, AnyRouterOptions>;

  const {
    state,
    descriptors,
    NavigationContent: Content,
  } = useNavigationBuilder<
    AnyNavigationState,
    AnyRouterOptions,
    AnyActionHelpers,
    TestScreenOptions,
    NavigationHelpers<ParamListBase>
  >(router, rest);

  return <Content>{state.routes.map((route) => descriptors[route.key]?.render())}</Content>;
};

/**
 * Creates a navigator factory for the TestNavigator.
 * This wraps the custom `TestNavigator` using `createNavigatorFactory`,
 * enabling use of `Navigator` and `Screen` components just like built-in navigators.
 *
 * Use this when integrating `TestNavigator` into `NavigationContainer`.
 */
const createTestNavigatorFactory = createNavigatorFactory(TestNavigatorImpl);

type LayoutProps = {
  name: string;
  focused: boolean;
  navigation: NavigationHelpers<ParamListBase>;
  children: React.ReactNode;
};

const CustomLayout = ({ name, focused, navigation, children }: LayoutProps): React.ReactElement => {
  useEffect(() => {
    const state = navigation.getState();
    const currentRoute = state.routes[state.index];

    if (focused && currentRoute?.name !== name) {
      navigation.navigate(name);
    } else if (!focused && currentRoute?.name !== 'Dummy') {
      navigation.navigate('Dummy');
    }
  }, [focused, name, navigation]);

  return <>{children}</>;
};

const TestLayout = ({ name, focused, ...props }: LayoutProps): React.ReactElement => {
  return <CustomLayout {...props} name={name} focused={focused} />;
};

export const TestNavigator = ({
  children,
  type = 'stack',
  name = 'TestScreen',
  focused = true,
  screenParams,
  dummyName = 'Dummy',
}: {
  children: React.ReactNode;
  type?: NavigatorType;
  name?: string;
  focused?: boolean;
  screenParams?: object;
  dummyName?: string;
}) => {
  const Test = createTestNavigatorFactory();

  return (
    <NavigationContainer>
      <Test.Navigator
        key={type}
        type={type}
        initialRouteName={focused ? name : dummyName}
        layout={
          /* eslint-disable-next-line react/no-unstable-nested-components */
          (props: Omit<LayoutProps, 'name' | 'focused'>) => (
            <TestLayout {...props} name={name} focused={focused} />
          )
        }
      >
        <Test.Screen name={name} getId={() => name} initialParams={screenParams}>
          {() => children}
        </Test.Screen>
        <Test.Screen name={dummyName} getId={() => dummyName}>
          {() => <Text testID="dummy-screen">Dummy</Text>}
        </Test.Screen>
      </Test.Navigator>
    </NavigationContainer>
  );
};
