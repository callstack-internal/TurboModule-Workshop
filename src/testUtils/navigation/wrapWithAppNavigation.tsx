import React from 'react';

import {
  InitialState,
  NavigationContainer,
  NavigationContainerRef,
} from '@react-navigation/native';
import DrawerNavigator from '../../navigation/navigators/DrawerNavigator.tsx';

type WrapWithAppNavigationProps = {
  initialState?: InitialState;
  navigationRef?: React.RefObject<NavigationContainerRef<any>>;
};

/**
 * Wraps the full app navigator (`DrawerNavigator`) in a `NavigationContainer` for testing.
 *
 * This is the preferred utility when writing tests for **real screen components or components that navigate** that are part of
 * the actual app navigation tree. It gives you a full, realistic navigation environment without mocks,
 * including nested navigators, screen options, transitions, etc.
 *
 * Unlike `TestNavigator`, this uses your actual app's navigator configuration, making it ideal for testing:
 * - Navigation behavior (e.g., `navigate`, `goBack`, deep linking)
 * - Screen lifecycle hooks (e.g., `useFocusEffect`, `useNavigation`)
 * - Integration-level rendering of screen flows
 *
 * ⚠️ Note: This should be used when testing **screen components**.
 * If you're only testing a non-screen component (e.g. custom hook, modal, header)
 * that requires navigation context, consider using `TestNavigator` or `createTestNavigator` instead for a minimal setup.
 *
 */
export const wrapWithAppNavigation = ({
  initialState,
  navigationRef,
}: WrapWithAppNavigationProps) => {
  return (
    <NavigationContainer initialState={initialState} ref={navigationRef}>
      <DrawerNavigator />
    </NavigationContainer>
  );
};
