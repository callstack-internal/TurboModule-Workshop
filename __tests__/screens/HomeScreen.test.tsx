import { render, waitFor } from '@testing-library/react-native';
import { Screens } from '../../src/navigation/constants.ts';
import { wrapWithAppNavigation } from '../../src/testUtils/navigation/wrapWithAppNavigation.tsx';
import { TestNavigator } from '../../src/testUtils/navigation/TestNavigator.tsx';
import HomeScreen from '../../src/screens/HomeScreen.tsx';

const initialState = {
  routes: [
    {
      name: Screens.Main.name,
      state: {
        routes: [
          {
            name: Screens.Home.name,
            state: {
              index: 1,
              routes: [
                {
                  name: Screens.Adoptar.name,
                },
              ],
            },
          },
        ],
      },
    },
  ],
};

const component = () => render(wrapWithAppNavigation({ initialState }));

describe('HomeScreen', () => {
  test('renders correctly', async () => {
    const { getByTestId, findByText } = component();

    await waitFor(() => {
      expect(getByTestId('home-screen-container')).toBeTruthy();
      expect(findByText('CatÃ¡logo de Perros')).toBeTruthy();
    });
  });

  test('matches snapshot', async () => {
    const { toJSON } = render(
      <TestNavigator>
        <HomeScreen />
      </TestNavigator>,
    );
    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
