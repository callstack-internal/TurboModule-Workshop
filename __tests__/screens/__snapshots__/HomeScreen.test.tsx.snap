// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen matches snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#f8f9fa",
      "flex": 1,
    }
  }
  testID="home-screen-container"
>
  <View
    style={
      {
        "backgroundColor": "#f8f9fa",
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "#ffffff",
          "borderBottomColor": "#e9ecef",
          "borderBottomWidth": 1,
          "elevation": 2,
          "padding": 16,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 3,
        }
      }
    >
      <Text
        style={
          {
            "color": "#212529",
            "fontSize": 24,
            "fontWeight": "bold",
            "marginBottom": 12,
          }
        }
      >
        Catálogo de Perros
      </Text>
      <TextInput
        onChangeText={[Function]}
        placeholder="Buscar por nombre"
        placeholderTextColor="#6c757d"
        style={
          {
            "backgroundColor": "#ffffff",
            "borderColor": "#ced4da",
            "borderRadius": 8,
            "borderWidth": 1,
            "color": "#212529",
            "fontSize": 16,
            "height": 46,
            "marginBottom": 16,
            "paddingHorizontal": 16,
          }
        }
        value=""
      />
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginBottom": 12,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#0066cc",
              "borderRadius": 8,
              "flex": 1,
              "justifyContent": "center",
              "marginRight": 8,
              "opacity": 1,
              "paddingHorizontal": 16,
              "paddingVertical": 10,
            }
          }
        >
          <Text
            style={
              {
                "color": "#ffffff",
                "fontSize": 14,
                "fontWeight": "600",
              }
            }
          >
            Refrescar Perros
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#28a745",
              "borderRadius": 8,
              "flex": 1,
              "justifyContent": "center",
              "marginLeft": 8,
              "marginRight": 0,
              "opacity": 1,
              "paddingHorizontal": 16,
              "paddingVertical": 10,
            }
          }
        >
          <Text
            style={
              {
                "color": "#ffffff",
                "fontSize": 14,
                "fontWeight": "600",
              }
            }
          >
            Obtener Ubicación
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <ActivityIndicator
        color="#0066cc"
        size="large"
      />
      <Text
        style={
          {
            "color": "#6c757d",
            "fontSize": 16,
            "marginTop": 12,
          }
        }
      >
        Cargando perros...
      </Text>
    </View>
  </View>
</View>
`;
